Q1: 
Are the majority of the CpG dinucleotides methylated or unmethylated? 
The majority seem to be methylated.

Q3a:
Bisulfite unique reads: 132466 (2.9908053720974968) %
Bisulfite multi reads: 4296642 (97.0091946279025) %
ONT unique reads: 53346 (1.2263482106157535) %
ONT multi reads: 4296642 (98.77365178938425) %
Total multi: 4296642 (95.85468138657977) %

Q2:
How does using nanopore for methylation calling differ from bisulfite sequencing in terms of coverage? Which method appears better and why?
Nanopore sequencing results in more consistent coverage than bisulfite (coverage distribution is a more narrow peak), however the coverage at this peak is lower than in bisulfite. Thus, since bisulfite sequencing gives a higher average coverage, the bisulfite sequencing is better.

Q3:
What can you infer about the two different approaches and their ability to detect methylation changes? 
While both approaches show similar overall detection of changes in methylation, the variation between the scores is significantly higher in the nanopore sequencing. Thus, bisulfite sequencing appears to be more consistent in its detection of methylation changes.


Q4: What is the effect of tumorigenesis on global methylation patterns?
Tumorigenesis results (mostly) in an increase in global methylation patters, though some loci are demethylated. This makes sense, as in order to grow, tumors need to suppress the expression of many growth-regulatory and tumor-supressor genes, while activating some that allow it to redirect blood flow and gain new abilities.